File: Java/Language/Generics/TypeParamTest1/Program3.java
class Program {

	//a generic method declaration with type-parameter T
	private static <T> T select(int index, T first, T second) {
		if((index % 2) == 1)
			return first;
		return second;
	}

	//a generic method declaration with type-parameter T
	//bounded by java.lang.Comparable<T> to extend set of 
	//members supported by T with the members of java.lang.Comparable<T>
	private static <T extends Comparable<T>> T select(T first, T second){
		if(first.compareTo(second) > 0)
			return first;
		return second;
	}

	public static void main(String[] args) {
		if(args.length > 0){
			int s = Integer.parseInt(args[0]);
			String ss = select(s, "Monday", "Friday");
			System.out.printf("Selected String = %s%n", ss);
			//auto-boxing - enclosing a value of primitive type (double) 
			//within an instance of its wrapper class (java.lang.Double)
			//so that it can be passed in place of java.lang.Object
			double sd = select(s, 45.6, 76.5);
			System.out.printf("Selected double = %s%n", sd);			
			//double sd2 = select(s, "9.5", 5.9);
		}else{
			String ss = select("Monday", "Friday");
			System.out.printf("Selected String = %s%n", ss);
			double sd = select(45.6, 76.5);
			System.out.printf("Selected double = %s%n", sd);
			Interval si = select(new Interval(4, 50), new Interval(3, 45));
			System.out.printf("Selected Interval = %s%n", si);
		}
	}
}

File: dotnet/Language/Generics/TypeParamTest1/DemoApp/Program3.cs
class Program
{
    //a generic method declaration with type-parameter T
    private static T Select<T>(int index, T first, T second)
    {
        if((index % 2) == 1)
            return first;
        return second;
    }

    private static T Select<T>(T first, T second) where T: IComparable<T>
    {
        //if(first.GetHashCode() > second.GetHashCode())
        if(first.CompareTo(second) > 0)
            return first;
        return second;
            
    }
    
    public static void Main(string[] args)
    {
        if(args.Length > 0)
        {
            int s = int.Parse(args[0]);
            string ss = Select(s, "Monday", "Tuesday");
            Console.WriteLine($"Selected string = {ss}");
            double sd = Select(s, 45.6, 34.5);
            Console.WriteLine($"Selected double = {sd}");
            //double ssd = Select(s, "Monday", 34.5);
        }
        else
        {
            string ss = Select("Monday", "Tuesday");
            Console.WriteLine($"Selected string = {ss}");
            double sd = Select(45.6, 34.5);
            Console.WriteLine($"Selected double = {sd}");   
            Interval si = Select(new Interval(3, 45), new Interval(4, 30));       
            Console.WriteLine($"Selected Interval = {si}");
        }
    }
}